semhash/semhash.py

```diff
@@ -6,139 +6,162 @@
-
-        # Records is a list of strings
-        embeddings = self.model.encode(records)
-        return embeddings
-
-    def _unpack_record(self, record: Record) -> str:
-        r"""
-        Unpack a record into a single string.
-
-        If the record is a dictionary, it uses self.columns to determine the order of the text segments.
-        Each text is cleaned by replacing '\t' with ' '. The texts are then joined by '\t'.
-
-        If the record is a string, it just replaces '\t' with ' ' and returns it.
         """
-        if isinstance(record, dict):
-            if self.columns is None:
-                raise ValueError("Columns must be specified when passing dictionaries.")
-
-            column_texts = []
-            for column in self.columns:
-                text = record[column].replace("\t", " ")
-                column_texts.append(text)
-            return "\t".join(column_texts)
+        # Extract the embeddings for each column across all records
+        embeddings_per_col = []
+        for col in columns:
+            col_texts = [r[col] for r in records]
+            col_emb = model.encode(col_texts)
+            embeddings_per_col.append(np.asarray(col_emb))
 
-        # Record is a string
-        return record.replace("\t", " ")
+        return np.concatenate(embeddings_per_col, axis=1)
 
+    @staticmethod
     def _remove_exact_duplicates(
-        self, records: Sequence[Record], reference_records: list[str] | None = None
-    ) -> tuple[list[Record], list[Record]]:
+        records: Sequence[RecordDict],
+        columns: Sequence[str],
+        reference_records: list[str] | None = None,
+    ) -> tuple[list[RecordDict], list[RecordDict]]:
         """
         Remove exact duplicates based on the unpacked string representation of each record.
 
         If reference_records is None, the function will only check for duplicates within the records list.
 
         :param records: A list of records to check for exact duplicates.
+        :param columns: Columns to unpack.
         :param reference_records: A list of records to compare against. These are already unpacked
         :return: A list of deduplicated records and a list of duplicates.
         """
         deduplicated = []
         duplicates = []
-        seen = set()
-        in_one_set = True
-        if reference_records is not None:
-            # If the reference records are provided, we need to add them to the seen set
-            # to deduplicate within the reference set
-            seen = set(reference_records)
-            in_one_set = False
+
+        # Build a seen set from reference_records if provided
+        seen = set(reference_records) if reference_records else set()
+        in_one_set = reference_records is None
 
         for record in records:
-            unpacked_record = self._unpack_record(record)
+            unpacked_record = SemHash._unpack_record(record, columns)
```

Comment:
This is dangerous. IMO, this should be a classmethod. For example, if you were to subclass `SemHash`, and would override `_unpack_record`, this function would still refer to the `_unpack_record` of the super class, leading to unexpected behavior.